src/librustc/mir/interpret/allocation.rs

```diff
@@ -41,6 +45,481 @@ pub struct Allocation<Tag=(),Extra=()> {
+                }
+                None => {},
+            }
+        }
+        // We don't. Just return the bits.
+        Ok(ScalarMaybeUndef::Scalar(Scalar::from_uint(bits, size)))
+    }
+
+    pub fn read_ptr_sized(
+        &self,
+        cx: &impl HasDataLayout,
+        ptr: Pointer<Tag>,
+        ptr_align: Align
+    ) -> EvalResult<'tcx, ScalarMaybeUndef<Tag>> {
+        self.read_scalar(cx, ptr, ptr_align, cx.data_layout().pointer_size)
+    }
+
+    /// Write a *non-ZST* scalar
+    pub fn write_scalar(
+        &mut self,
+        cx: &impl HasDataLayout,
+        ptr: Pointer<Tag>,
+        ptr_align: Align,
+        val: ScalarMaybeUndef<Tag>,
+        type_size: Size,
+    ) -> EvalResult<'tcx> {
+        let val = match val {
+            ScalarMaybeUndef::Scalar(scalar) => scalar,
+            ScalarMaybeUndef::Undef => return self.mark_definedness(ptr, type_size, false),
+        };
+
+        let bytes = match val {
+            Scalar::Ptr(val) => {
+                assert_eq!(type_size, cx.data_layout().pointer_size);
+                val.offset.bytes() as u128
+            }
+
+            Scalar::Bits { bits, size } => {
+                assert_eq!(size as u64, type_size.bytes());
+                debug_assert_eq!(truncate(bits, Size::from_bytes(size.into())), bits,
+                    "Unexpected value of size {} when writing to memory", size);
+                bits
+            },
+        };
+
+        {
+            // get_bytes_mut checks alignment
+            let endian = cx.data_layout().endian;
+            let dst = self.get_bytes_mut(cx, ptr, type_size, ptr_align)?;
+            write_target_uint(endian, dst, bytes).unwrap();
+        }
+
+        // See if we have to also write a relocation
+        match val {
+            Scalar::Ptr(val) => {
+                self.relocations.insert(
+                    ptr.offset,
+                    (val.tag, val.alloc_id),
+                );
+            }
+            _ => {}
+        }
+
+        Ok(())
+    }
+
+    pub fn write_ptr_sized(
+        &mut self,
+        cx: &impl HasDataLayout,
+        ptr: Pointer<Tag>,
+        ptr_align: Align,
+        val: ScalarMaybeUndef<Tag>
+    ) -> EvalResult<'tcx> {
+        let ptr_size = cx.data_layout().pointer_size;
+        self.write_scalar(cx, ptr.into(), ptr_align, val, ptr_size)
+    }
+
+    fn int_align(
+        &self,
+        cx: &impl HasDataLayout,
+        size: Size,
+    ) -> Align {
+        // We assume pointer-sized integers have the same alignment as pointers.
+        // We also assume signed and unsigned integers of the same size have the same alignment.
+        let ity = match size.bytes() {
+            1 => layout::I8,
+            2 => layout::I16,
+            4 => layout::I32,
+            8 => layout::I64,
+            16 => layout::I128,
+            _ => bug!("bad integer size: {}", size.bytes()),
+        };
+        ity.align(cx)
+    }
+}
+
+/// Byte accessors
```

Comment:
Byte accessors used to be above the methods that use it. Why reorder?

Seems like a good idea to roughly have the blocks in order of increasing "level".